src/openforms/payments/tests/test_views.py

```diff
@@ -205,3 +212,148 @@ def test_start_plugin_not_enabled(
+    CORS_ALLOW_ALL_ORIGINS=False,
+    CORS_ALLOWED_ORIGINS=["http://allowed.foo"],
+)
+class WebhookReturnTests(TestCase):
+    @classmethod
+    def setUpTestData(cls):
+        super().setUpTestData()
+
+        register = Registry()
+        register("payment-plugin")(PaymentPlugin)
+        cls.plugin = register["payment-plugin"]
+        cls.register = register
+
+    def setUp(self):
+        super().setUp()
+
+        registry_mock = patch("openforms.payments.views.register", new=self.register)
+        registry_mock.start()
+
+        self.addCleanup(registry_mock.stop)
+
+    @tag("gh-4052")
+    def test_payment_cancelled_does_not_trigger_on_completion_in_webhook_endpoint(self):
+        """
+        The webhook can be called also in the case that the payment was cancelled/failed.
+        In this case, we don't want to trigger the task:
+
+            on_post_submission_event(
+                <payment ID>, PostSubmissionEvents.on_payment_complete
+            )
+        """
+        submission = SubmissionFactory.create(
+            with_public_registration_reference=True,
+            form__product__price=Decimal("11.25"),
+            form__payment_backend="payment-plugin",
+            form_url="http://allowed.foo/my-form",
+        )
+        payment = SubmissionPaymentFactory.for_submission(
+            submission, status=PaymentStatus.failed
+        )
+
+        base_request = RequestFactory().get("/foo")
+
+        webhook_url = self.plugin.get_webhook_url(base_request)
+
+        with self.captureOnCommitCallbacks(execute=True) as callbacks:
+            response = self.client.post(
+                webhook_url, data={"orderID": payment.public_order_id}
+            )
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        # Payment status is 'failed', so no 'transaction.on_commit' call that triggers the on_post_submission_event
+        self.assertEqual(len(callbacks), 0)
+
+    def test_successful_payment_triggers_on_completion_in_webhook_endpoint(self):
+        submission = SubmissionFactory.create(
+            with_public_registration_reference=True,
+            form__product__price=Decimal("11.25"),
+            form__payment_backend="payment-plugin",
+            form_url="http://allowed.foo/my-form",
+        )
+        payment = SubmissionPaymentFactory.for_submission(
+            submission, status=PaymentStatus.completed
+        )
+
+        base_request = RequestFactory().get("/foo")
+
+        webhook_url = self.plugin.get_webhook_url(base_request)
+
+        with self.captureOnCommitCallbacks(execute=True) as callbacks:
+            response = self.client.post(
+                webhook_url, data={"orderID": payment.public_order_id}
+            )
+
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        # Payment was successful, so 'transaction.on_commit' is called and triggers the on_post_submission_event
+        self.assertEqual(len(callbacks), 1)
```

Comment:
I have some mixed feelings about these assertions - they will break as soon as another `on_commit` handler is added anywhere in the stack (so it could be a call downstream, or even a call via middleware that adds an on_commit handler completely unrelated to the problem at hand here).

I think I'd rather opt for a `patch("openforms.payments.views.on_post_submission_event")` followed by a `mock_on_post_submission_event.assert_called_once_with(submission.pk, PostSubmissionEvents.on_payment_complete)`, and for the test above an assertion that it's *not* called.